core/src/main/java/com/datastax/oss/driver/internal/core/metadata/DefaultTabletMap.java

```diff
@@ -0,0 +1,267 @@
+    for (HostShardPair hostShardPair : row.getReplicas()) {
+      result.add(hostShardPair.getHost());
+    }
+    return result;
+  }
+
+  /**
+   * Updates tablet map with parsed tablets-routing-v1 custom payload. Expects the tuple's
+   * underlying structure to correspond to {@code TupleType(LongType, LongType,
+   * ListType(TupleType(UUIDType, Int32Type)))}. Handles removing outdated tables that intersect
+   * with the one about to be added.
+   *
+   * @param keyspace the keyspace of the table
+   * @param table the table name
+   * @param tupleValue the parsed payload
+   */
+  @Override
+  public void addTablet(String keyspace, String table, TupleValue tupleValue) {
+    KeyspaceTableNamePair ktPair = new KeyspaceTableNamePair(keyspace, table);
+
+    long firstToken = tupleValue.getLong(0);
+    long lastToken = tupleValue.getLong(1);
+
+    logger.trace(
+        "Received tablets routing V1 payload: {}.{} range {}-{}",
+        keyspace,
+        table,
+        firstToken,
+        lastToken);
+
+    Set<HostShardPair> replicas = new HashSet<>();
+    List<TupleValue> list = tupleValue.getList(2, TupleValue.class);
+    assert list != null;
+    for (TupleValue tuple : list) {
+      HostShardPair hostShardPair = new HostShardPair(tuple.getUuid(0), tuple.getInt(1));
+      replicas.add(hostShardPair);
+    }
+
+    // Working on a copy to avoid concurrent modification of the same set
+    NavigableSet<Tablet> existingTablets =
+        new TreeSet<>(mapping.computeIfAbsent(ktPair, k -> new TreeSet<>()));
+
+    // Single tablet token range is represented by (firstToken, lastToken] interval
+    // We need to do two sweeps: remove overlapping tablets by looking at lastToken of existing
+    // tablets
+    // and then by looking at firstToken of existing tablets. Currently, the tablets are sorted
+    // according
+    // to their lastTokens.
+
+    // First sweep: remove all tablets whose lastToken is inside this interval
+    Iterator<Tablet> it =
+        existingTablets
+            .headSet(DefaultTablet.malformedTablet(lastToken), true)
+            .descendingIterator();
+    while (it.hasNext()) {
+      Tablet tablet = it.next();
+      if (tablet.getLastToken() <= firstToken) {
+        break;
+      }
+      it.remove();
+    }
+
+    // Second sweep: remove all tablets whose firstToken is inside this tuple's (firstToken,
+    // lastToken]
+    // After the first sweep, this theoretically should remove at most 1 tablet
+    it = existingTablets.tailSet(DefaultTablet.malformedTablet(lastToken), true).iterator();
+    while (it.hasNext()) {
+      Tablet tablet = it.next();
+      if (tablet.getFirstToken() >= lastToken) {
+        break;
+      }
+      it.remove();
+    }
+
+    // Add new (now) non-overlapping tablet
+    existingTablets.add(new DefaultTablet(keyspace, null, table, firstToken, lastToken, replicas));
+
+    // Set the updated result in the main map
+    mapping.put(ktPair, existingTablets);
```

Comment:
You said that working on a copy prevents concurrent modifications, but it looks like it introduces other serious problem. This whole method:
1. Clones mapping
2. Modifes mapping
3. Replaces mapping

What happends if 2 calls are made concurrently? Both will copy the field, calculate new mapping and then replace it - one of updates will be lost. Considering that tablet update is not a cheap operation and that there might be a lot of them after session initialization (making this race very likely) it doesn't seem like a good solution.

Reply:
This version was bugged, but a revised one makes modifications sequentially. I've made it so MetadataManager handles those modifications like other metadata changes. This has some room for improvement but it's probably good enough for now.